File: Typescript/MatrixCalculation.ts
function MatrixProduct(l:number[][], L:number[][]) {
  var res:number[][] = [];
  for (var i:number = 0; i<l.length; i++) {
    var a:number[] = [];
    for (var j:number = 0; j<L[0].length; j++) {
      a.push(Array.from(new Array(l.length).keys()).map(e=>l[i][e]*L[e][j]).reduce((a:number,b:number)=>a+b))
    }
    res.push(a)
  }
  return res
}

function Inverse(n:number[][]) {
  var o = n.length;
  function y(r:number[][]) {
    var Q:number[][] = [];
    for (var i = 0; i < o; i++) {
      var R:number[] = [];
      for (var j = 0; j < o; j++) R.push(r[j][i]);
      Q.push(R)
    }
    return Q
  }
  function a(s:number[][]) {
    var w = s.length;
    if (w == 1) return s[0][0];
    if (w == 2) return s[0][0] * s[1][1] - s[1][0] * s[0][1];
    else {
      var res = 0;
      for (var i = 0; i < w; i++) res += s[0][i] * a(c(s, 0, i)) * Math.pow(-1, i);
      return res
    }
  }
  function c(q:number[][], I:number, J:number) {
    return q.slice(0, I).concat(q.slice(I + 1, q.length)).map(e => e.slice(0, J).concat(e.slice(J + 1, q.length)));
  }
  var D = 1 / a(n);
  var p:number[][] = [];
  for (var i = 0; i < o; i++) {
    var T:number[] = [];
    for (var j = 0; j < o; j++) T.push(Math.pow(-1, i + j) * D * a(c(n, i, j)));
    p.push(T)
  }
  return y(p)
}

File: Lua/MatrixCalculation.lua
function MatrixProduct(a,b)
  res = {}
  for i=1, #a do
    p = {}
    for j=1, #(b[1]) do
      r = 0
      for k=1, #(a[1]) do
        r = r + a[i][k]*b[k][j]
      end
      table.insert(p,r)
    end
    table.insert(res,p)
  end
  return res
end

function Inverse(W)
  o = #W
  function Transpose(v)
    res = {}
    for i=1, o do
      R = {}
      for j=1, o do table.insert(R,v[j][i])
      end
      table.insert(res,R)
    end
    return res
  end

  function Cofactor(v,I,J)
    function copy1(obj)
      if type(obj) ~= 'table' then return obj end
      local res = {}
      for k, v in pairs(obj) do res[copy1(k)] = copy1(v) end
      return res
    end

    q = copy1(v)
    table.remove(q,I)
    for index, i in pairs(q) do table.remove(i,J) end
    return q
  end

  function Abs(m)
    w = #m
    if w == 1 then return m[1][1]
    elseif w == 2 then return m[1][1]*m[2][2] - m[1][2]*m[2][1]
    else
      res = 0
      for i=1, w do
        res = res + m[1][i]*Abs(Cofactor(m,1,i))*math.pow(-1,i)
      end
      return res
    end
  end

  D = 1/Abs(W)
  Res = {}
  for i=1, o do
    t = {}
    for j=1, o do
      table.insert(t, math.pow(-1,i+j)*D*Abs(Cofactor(W,i,j)))
    end
    table.insert(Res,t)
  end
  return Transpose(Res)
end
